### PR TITLE
Gan training

### PR DESCRIPTION
**Pull request by MylesBartlett**

### CODE REVIEW

File: main_gan.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+                        help='evaluate model on validation set (default: false)')
+
+    # gpu/cpu
+    parser.add_argument('--gpu_num', type=int, default=0, metavar='GPU',
+                        help='choose GPU to run on.')
+
+    # CNN
+    parser.add_argument('--cnn_in_channels', type=int, default=3)
+    parser.add_argument('--cnn_hidden_channels', type=int, default=32)
+    parser.add_argument('--cnn_num_hidden_layers', type=int, default=7)
```
Comment (tmke8) [inline main_gan.py]: I think you have to change it here now to `cnn_hidden_layers`

Comment (MylesBartlett) [reply main_gan.py]: Just did that, anticipating your comment :)

File: models/simple_complex_cnn.py [hunk 2]
```diff
@@ -19,7 +19,7 @@ def __init__(self, args):
                                         local_condition=args.iso)]
         # Hidden layers
         for _ in range(args.cnn_num_hidden_layers):
-            layers.append(ComplexGatedConvLayer(args.cnn_hidden_channels, args.cnn_hidden_channels,
+            layers.append(ComplexGatedConvLayer(num_hidden_channels, num_hidden_channels,
```
Comment (tmke8) [inline models/simple_complex_cnn.py]: This change is causing a conflict because this line was also changed on master.

Comment (MylesBartlett) [reply models/simple_complex_cnn.py]: The hidden layers should take the local variable 'num_hidden_channels' as the argument to their constructor. I only have access to my phone atm with which it's not easy to resolve conflicts due to the formatting.

Comment (tmke8) [reply models/simple_complex_cnn.py]: Ok, I can resolve the conflict

Comment (MylesBartlett) [reply models/simple_complex_cnn.py]: Thanks

**Comment by tmke8** [comment]
Is it correct that in every iteration of the for loop in simple_complex_cnn.py **2** ComplexGatedConvLayer are appended to the list?

File: models/simple_complex_cnn.py (hunk 2 above)
Comment (tmke8) [reply models/simple_complex_cnn.py]: done

**Comment by MylesBartlett** [comment]
Ah, that's a mistake. Should just be one layer per iteration.

**Comment by MylesBartlett** [comment]
Incidentally, I think we should remove 'Simple' from the class names of the gated and complex models to make them less verbose.

File: main_gan.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+from optimisation.testing import test
+from optimisation.training import train, train_gan, validate, evaluate_psnr_and_vgg_loss
+from optimisation import loss
+from utils import TransformedHuaweiDataset, transform_sample
+from utils.functions import apply_spectral_norm
+import models
+
+
+def parse_arguments(raw_args=None):
```
Comment (tmke8) [inline main_gan.py:20]: All the code duplication is not nice (makes it very hard to maintain). Can you at least import this function from main.py? I think it should be the same there.

Comment (MylesBartlett) [reply main_gan.py:20]: The arguments for main_gan.py are different.

**Pull request merged**